File: cdm-sources/cdm/product/qualification/functions/UnderlierQualification.java
package cdm.product.qualification.functions;

import cdm.base.staticdata.asset.common.AssetClassEnum;
import cdm.base.staticdata.asset.common.Instrument;
import cdm.base.staticdata.asset.common.InstrumentTypeEnum;
import cdm.base.staticdata.asset.common.Security;
import cdm.observable.asset.metafields.ReferenceWithMetaObservable;
import cdm.product.template.Product;
import cdm.product.template.TransferableProduct;
import cdm.product.template.Underlier;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.expression.CardinalityOperator;
import com.rosetta.model.lib.expression.ComparisonResult;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperS;
import javax.inject.Inject;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(UnderlierQualification.UnderlierQualificationDefault.class)
public abstract class UnderlierQualification implements RosettaFunction {
	
	// RosettaFunction dependencies
	//
	@Inject protected ObservableQualification observableQualification;

	/**
	* @param underlier An Underlier is an Observable (eg Asset, Basket or Index) or a Product.
	* @param securityType 
	* @param assetClass 
	* @return qualifies 
	*/
	public Boolean evaluate(Underlier underlier, InstrumentTypeEnum securityType, AssetClassEnum assetClass) {
		Boolean qualifies = doEvaluate(underlier, securityType, assetClass);
		
		return qualifies;
	}

	protected abstract Boolean doEvaluate(Underlier underlier, InstrumentTypeEnum securityType, AssetClassEnum assetClass);

	public static class UnderlierQualificationDefault extends UnderlierQualification {
		@Override
		protected Boolean doEvaluate(Underlier underlier, InstrumentTypeEnum securityType, AssetClassEnum assetClass) {
			Boolean qualifies = null;
			return assignOutput(qualifies, underlier, securityType, assetClass);
		}
		
		protected Boolean assignOutput(Boolean qualifies, Underlier underlier, InstrumentTypeEnum securityType, AssetClassEnum assetClass) {
			final ReferenceWithMetaObservable referenceWithMetaObservable = MapperS.of(underlier).<ReferenceWithMetaObservable>map("getObservable", _underlier -> _underlier.getObservable()).get();
			qualifies = ComparisonResult.of(MapperS.of(observableQualification.evaluate((referenceWithMetaObservable == null ? null : referenceWithMetaObservable.getValue()), securityType, assetClass))).or(areEqual(MapperS.of(underlier).<Product>map("getProduct", _underlier -> _underlier.getProduct()).<TransferableProduct>map("getTransferableProduct", product -> product.getTransferableProduct()).<Instrument>map("getInstrument", transferableProduct -> transferableProduct.getInstrument()).<Security>map("getSecurity", instrument -> instrument.getSecurity()).<InstrumentTypeEnum>map("getInstrumentType", security -> security.getInstrumentType()), MapperS.of(securityType), CardinalityOperator.All)).get();
			
			return qualifies;
		}
	}
}

File: cdm-sources/src/cdm/product/qualification/functions/UnderlierQualification.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.base.staticdata.asset.common.AssetClassEnum import AssetClassEnum
from cdm.product.template.Underlier import Underlier
from cdm.product.qualification.functions.ObservableQualification import ObservableQualification
from cdm.base.staticdata.asset.common.InstrumentTypeEnum import InstrumentTypeEnum

__all__ = ['UnderlierQualification']


@replaceable
def UnderlierQualification(underlier: Underlier, securityType: InstrumentTypeEnum | None, assetClass: AssetClassEnum | None) -> bool:
    """
    Identifies whether the underlier(s) have either the specified securityType or assetClass.
    
    Parameters 
    ----------
    underlier : Underlier
    An Underlier is an Observable (eg Asset, Basket or Index) or a Product.
    
    securityType : InstrumentTypeEnum
    
    assetClass : AssetClassEnum
    
    Returns
    -------
    qualifies : boolean
    
    """
    self = inspect.currentframe()
    
    
    qualifies =  (ObservableQualification(rosetta_resolve_attr(rosetta_resolve_attr(self, "underlier"), "Observable"), rosetta_resolve_attr(self, "securityType"), rosetta_resolve_attr(self, "assetClass")) or all_elements(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "underlier"), "Product"), "TransferableProduct"), "Instrument"), "Security"), "instrumentType"), "=", rosetta_resolve_attr(self, "securityType")))
    
    
    return qualifies

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)
